### ISSUE TITLE
useless_asref dosent understand ` &mut &'a`.

### ISSUE THREAD

**Issue by pigeonhands** [rust-lang/rust-clippy#10790]
### Summary

Clippy is saying to remove the "`useless_asref`" even though it does not work without it.

### Lint Name

useless_asref

### Reproducer

I tried this code:

```rust
fn test_fn<'a>(data: &mut &'a [u8], take: usize) -> &'a [u8] {
    let (val, remaining) = data.split_at(take);
        *data = remaining;
        val
}

fn main(){
    let data : &[u8] = &[0u8,1,2,3,4,5];
    test_fn(&mut data.as_ref(), 2);
}
```

I saw this happen:

```
this call to `as_ref` does nothing
```

I expected to see this happen:
Should not show lint as it does not work without it.

### Version

```text
rustc 1.68.2 (9eb3afe9e 2023-03-27)
binary: rustc
commit-hash: 9eb3afe9ebe9c7d2b84b71002d44f4a0edac95e0
commit-date: 2023-03-27
host: x86_64-pc-windows-msvc
release: 1.68.2
LLVM version: 15.0.6
```

**Comment by y21** [comment]
IMO the lint is right. The call to `as_ref` is not really necessary, since `data.as_ref()` is basically a no-op in this instance; it goes from `&[u8]` to `&[u8]`. Not a really useful conversion.
It *should* just work without it, but it doesn't for a different, unrelated reason. The `data` binding needs to be made mutable, and then it compiles ([see here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=dd2a435cf131eb8c311067cbfd784dc9)).

I believe the reason for why the `data` binding didn't need to be mutable *with* `as_ref` is that [Temporary Lifetime Extension](https://doc.rust-lang.org/reference/destructors.html?highlight=destructor#temporary-lifetime-extension) kicks in and gives the temporary returned by `as_ref` its own hidden mutable binding. That is, `test_fn(&mut data.as_ref(), 2)` becomes
```rs
let data = &[0, 1, 2, 3, 4, 5];
let mut hidden = data.as_ref();
test_fn(&mut hidden, 2);
```
... which is interesting, because that means that the code behaves differently depending on if `as_ref` is used or not used (with `as_ref` it mutates some hidden binding introduced by the compiler and without `as_ref` it mutates `data`).

Though I do agree that *something* is wrong here.
